core/finality-grandpa/src/lib.rs

```diff
@@ -0,0 +1,767 @@
+	}
+
+	fn best_chain_containing(&self, block: Block::Hash) -> Option<(Block::Hash, u32)> {
+		match self.inner.best_containing(block, None) {
+			Ok(Some(hash)) => {
+				let header = self.inner.header(&BlockId::Hash(hash)).ok()?
+					.expect("Header known to exist after `best_containing` call; qed");
+
+				Some((hash, header.number().as_()))
+			}
+			Ok(None) => None,
+			Err(e) => {
+				debug!(target: "afg", "Encountered error finding best chain containing {:?}: {:?}", block, e);
+				None
+			}
+		}
+	}
+}
+
+impl<B, E, Block: BlockT, N> voter::Environment<Block::Hash> for Environment<B, E, Block, N> where
+	Block: 'static,
+	B: Backend<Block, Blake2Hasher> + 'static,
+	E: CallExecutor<Block, Blake2Hasher> + 'static,
+	N: Network + 'static,
+	N::In: 'static,
+	NumberFor<Block>: As<u32>,
+{
+    type Timer = Box<Future<Item = (), Error = Self::Error>>;
+    type Id = AuthorityId;
+    type Signature = ed25519::Signature;
+    type In = Box<Stream<Item = ::grandpa::SignedMessage<Block::Hash, Self::Signature, Self::Id>, Error = Self::Error>>;
+    type Out = Box<Sink<SinkItem = ::grandpa::Message<Block::Hash>, SinkError = Self::Error>>;
+    type Error = Error;
+
+    fn round_data(
+        &self,
+        round: u64
+    ) -> voter::RoundData<Self::Timer, Self::Id, Self::In, Self::Out> {
+		use client::BlockchainEvents;
+		use tokio::timer::Delay;
+
+		let now = Instant::now();
+		let prevote_timer = Delay::new(now + self.config.gossip_duration * 2);
+		let precommit_timer = Delay::new(now + self.config.gossip_duration * 4);
+
+		// TODO: dispatch this with `mpsc::spawn`.
+		let incoming = checked_message_stream::<Block, _>(
+			self.network.messages_for(round),
+			self.config.voters.clone(),
+		);
+
+		let (out_rx, outgoing) = outgoing_messages::<Block, _>(
+			self.config.local_key.clone(),
+			self.config.voters.clone(),
+			round,
+			self.network.clone(),
+		);
+
+		// schedule incoming messages from the network to be held until
+		// corresponding blocks are imported.
+		let incoming = UntilImported::new(
+			self.inner.import_notification_stream(),
+			self.inner.clone(),
+			incoming,
+		);
+
+		// join incoming network messages with locally originating ones.
+		let incoming = Box::new(incoming.select(out_rx));
+
+		// schedule network message cleanup when sink drops.
+		let outgoing = Box::new(ClearOnDrop {
+			round,
+			network: self.network.clone(),
+			inner: outgoing,
+		});
+
+		voter::RoundData {
+			prevote_timer: Box::new(prevote_timer.map_err(Error::Timer)),
+    		precommit_timer: Box::new(precommit_timer.map_err(Error::Timer)),
+    		voters: self.voters.clone(),
+    		incoming,
+    		outgoing,
+		}
+	}
+
+    fn completed(&self, round: u64, state: RoundState<Block::Hash>) {
+		let encoded_state = (round, state).encode();
+		if let Err(e) = self.inner.backend()
+			.insert_aux(&[(LAST_COMPLETED_KEY, &encoded_state[..])], &[])
+		{
+			warn!(target: "afg", "Error bookkeeping last completed round in DB: {:?}", e);
```

Comment:
Error here is quite possible. Out of disk for example. Is simple warning enough?

Reply:
@rphmeier ^^^